feat(5388): Allow multiple external URLs for SSO access

ClifHouck opened the pull request:
My organization has interest in getting this functionality merged, so I'm taking up the mantle. I've re-based the original PR on current (79b18e03dc05f2d1da64d30650b12e064e6ef067) master. This currently builds, but tests do not pass. I'll be back online next week to fix tests, verify functionality of this change, and respond to any existing review comments from the original PR.

Preserving the original PR note here for reference:

With https://github.com/argoproj/argo-cd/pull/4780, were introduced security measures to ensure the return_url is pointing to the current ArgoCD instance.

In several occasions, an ArgoCD isntance could be exposed through multiple network connections. Internal addresses and restricted public addresses.

Currently, a single base URL can be configured in the the argocd configmap, preventing from exposing ArgoCD on several access paths.

This change allows to define multiple hosts on which ArgoCD can be exposed, keeping backward compatibility by adding a field additionalUrls accepting a list of additional URLS on which argoCD can be exposed

Fixes https://github.com/argoproj/argo-cd/issues/5388

<!--
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.
-->

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [x] Optional. My organization is added to USERS.md.
* [x] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

<!-- Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request. -->

agaudreault commented:
@ClifHouck make sure to adress the review in https://github.com/argoproj/argo-cd/pull/14208#pullrequestreview-2116512325. You can switch the PR to draft until the test are passing and everything is implemented. Thanks for taking up the mantle! 💪

ClifHouck commented on util/oidc/oidc.go:
```diff
@@ -148,17 +148,21 @@ func NewClientApp(settings *settings.ArgoCDSettings, dexServerAddr string, dexTl
 }
 
-func (a *ClientApp) oauth2Config(scopes []string) (*oauth2.Config, error) {
+func (a *ClientApp) oauth2Config(request *http.Request, scopes []string) (*oauth2.Config, error) {
```
It seems like the request object is needed here. Re: https://github.com/argoproj/argo-cd/pull/14208/files#r1638640243

ClifHouck commented on util/settings/settings.go:
```diff
@@ -1468,6 +1474,16 @@ func updateSettingsFromConfigMap(settings *ArgoCDSettings, argoCDCM *apiv1.Confi
 	}
+	if argoCDCM.Data[settingURLsKey] != "" {
+		if err := yaml.Unmarshal([]byte(argoCDCM.Data[settingURLsKey]), &settings.URLs); err != nil {
+			log.Warnf("Failed to decode all external URLs in configmap: %v", err)
```
Corrected messages here, re: https://github.com/argoproj/argo-cd/pull/14208/files#r1638642960

ClifHouck reviewed (commented):
Regarding this last comment: https://github.com/argoproj/argo-cd/pull/14208/files#r1638649274 it seems like Dex is not using the new URLs setting. So I removed that addition.

ClifHouck commented:
@agaudreault I think I've addressed the comments you had on the previous PR. Except this: 

> 
I think the feature should be documented in the ArgoCD https://argo-cd.readthedocs.io/en/stable/operator-manual/user-management/ documentation.

Agreed. I can add some documentation around this feature.

> Is url deprecated? are urls valid on their own or only additional urls? Is this feature supported by all identity providers, or only by OIDC?

@tjamet What do you think about the questions posed above?

ClifHouck commented:
What do I need to about the license compliance failing? Is there a contributor agreement I need to sign?

tjamet commented:
> @agaudreault I think I've addressed the comments you had on the previous PR. Except this: 
> 
> 
> 
> > I think the feature should be documented in the ArgoCD https://argo-cd.readthedocs.io/en/stable/operator-manual/user-management/ documentation.
> 
> 
> 
> Agreed. I can add some documentation around this feature.
> 
> 
> 
> > Is url deprecated? are urls valid on their own or only additional urls? Is this feature supported by all identity providers, or only by OIDC?
> 
> 
> 
> @tjamet What do you think about the questions posed above?

Hi!

Yes, it should be documented, happy to help doing so.

About the deprecation, this was my idea at the beginning but I faced some issues as `url` seemed used for more things. I will take a second look

ClifHouck commented:
@tjamet Any updates?

tjamet commented:
> @tjamet Any updates?

Taking a second look, indeed, URL is used to register in dex, and the update to replace by the new proposed list is not that trivial.

Eventually, a decision to deprecate the url config in favour of a list could be taken in the future

ClifHouck commented on docs/operator-manual/argocd-cm.yaml:
```diff
@@ -10,6 +10,10 @@ data:
   url: https://argo-cd-demo.argoproj.io
 
+  # Additional externally facing base URLs (optional)
+  urls: |
```
I think I've updated dex to at least load the additional URLs with the proper callback endpoint appended. 

I've also handled config reload when `settings.AdditionalURLs` is modified.

ClifHouck commented on util/oidc/oidc.go:
```diff
@@ -148,17 +148,21 @@ func NewClientApp(settings *settings.ArgoCDSettings, dexServerAddr string, dexTl
 	}
+	redirectURL, err := a.settings.RedirectURLForRequest(request)
+	if err != nil {
+		redirectURL = a.redirectURI
```
Added warning.

ClifHouck commented on docs/operator-manual/argocd-cm.yaml:
```diff
@@ -10,6 +10,10 @@ data:
   url: https://argo-cd-demo.argoproj.io
 
+  # Additional externally facing base URLs (optional)
+  urls: |
```
Renamed.

ClifHouck commented on docs/operator-manual/argocd-cm.yaml:
```diff
@@ -10,6 +10,10 @@ data:
+  # Additional externally facing base URLs (optional)
+  urls: |
+    - https://argo-cd-demo2.argoproj.io
+
```
Added some short documentation in the user-management doc

ClifHouck commented:
I believe I've addressed everything brought up so far, and re-based on latest master e2c5a555ce559f7284428a73565c53bbc9480986 for good measure. Not sure what to do about the linter failing on files I haven't changed.

ClifHouck commented:
Also need guidance on what to do about the license compliance check.

agaudreault commented:
>Also need guidance on what to do about the license compliance check.

@ClifHouck  You can ignore that.

agaudreault reviewed (commented):
LGTM. Added in my sprint to give it a try internally, for both dex and native oidc. I should be able to test tomorrow.

ClifHouck commented:
> LGTM. Added in my sprint to give it a try internally, for both dex and native oidc. I should be able to test tomorrow.

Thank you!

agaudreault commented on docs/operator-manual/argocd-cm.yaml:
```diff
@@ -10,6 +10,10 @@ data:
   url: https://argo-cd-demo.argoproj.io
 
+  # Additional externally facing base URLs (optional)
+  addtionalUrls: |
```
```suggestion
  additionalUrls: |
```

agaudreault commented:
@ClifHouck Have you tested this image on an actual deployment? I am getting constant `Invalid redirect URL: the protocol and host (including port) must match and the path must be within allowed URLs if provided`. Might be a configuration problem on my part, but everything looks as expected. I am testing with the native oidc auth.

agaudreault commented on util/settings/settings.go:
```diff
@@ -406,6 +410,8 @@ const (
 	settingServerPrivateKey = "tls.key"
 	// settingURLKey designates the key where Argo CD's external URL is set
 	settingURLKey = "url"
+	// settingURLsKey designates the key where Argo CD's external URLs is set
+	settingURLsKey = "urls"
```
Invalid key to load configuration

```suggestion
	// settingAdditionalUrlsKey designates the key where Argo CD's additional external URLs are set
	settingAdditionalUrlsKey = "additionalUrls"
```

ClifHouck commented:
> @ClifHouck Have you tested this image on an actual deployment? I am getting constant `Invalid redirect URL: the protocol and host (including port) must match and the path must be within allowed URLs if provided`. Might be a configuration problem on my part, but everything looks as expected. I am testing with the native oidc auth.

No, but I thought the included additional tests of the original PR would've covered this case. I'll setup an environment to do more in-depth testing. I may not be able to complete this before leaving for vacation though. I'll be back online on August 12th.

agaudreault reviewed (changes requested):
Using the correct value from config, it is working for dex+oidc and native oidc. 🎉

agaudreault commented on util/settings/settings.go:
```diff
@@ -1470,6 +1476,16 @@ func updateSettingsFromConfigMap(settings *ArgoCDSettings, argoCDCM *apiv1.Confi
 	if err := validateExternalURL(argoCDCM.Data[settingUiBannerURLKey]); err != nil {
 		log.Warnf("Failed to validate UI banner URL in configmap: %v", err)
 	}
+	if argoCDCM.Data[settingURLsKey] != "" {
+		if err := yaml.Unmarshal([]byte(argoCDCM.Data[settingURLsKey]), &settings.AdditionalURLs); err != nil {
+			log.Warnf("Failed to decode all external URLs in configmap: %v", err)
```
```suggestion
	if argoCDCM.Data[settingAdditionalUrlsKey] != "" {
		if err := yaml.Unmarshal([]byte(argoCDCM.Data[settingAdditionalUrlsKey]), &settings.AdditionalURLs); err != nil {
			log.Warnf("Failed to decode all additional URLs in configmap: %v", err)
```

agaudreault reviewed (approved):
Lgtm @ishitasequeira

ishitasequeira reviewed (approved):
Thanks @ClifHouck for the PR!! and Thanks @agaudreault for the review!!

alexmt reviewed (commented):
LGTM

akloss-cibo commented:
Any thoughts about when this will be in a tagged release?

dave-gantenbein commented:
@ishitasequeira - we could really use a tagged release that contains this change, any idea when that might happen?

blakepettersson commented:
This will be in `2.13`.
